### DISCUSSION TITLE
Runtime method to validate assignment when `validate_assignment=False` in config

### DISCUSSION THREAD

**Discussion by hassec** [pydantic/pydantic#7194] [Feature Request]
I'd like the option to opt in at runtime into validated assignment if the model was created with `validate_assignmetn=False`. 

As an example:
```py
from pydantic import BaseModel

class A(Basemodel):
    x : int

t = A(x=10)

# I can do
t.__pydantic_validator__.validate_assignment(t, 'x', '1')

# how about 
t.validate_assignment(...)

# or even better
with t.validated_assignement():
    t.x = 'a'

```

As shown I can use the `__pydantic_validator__` at runtime when I want validated assignment, but I'm not sure if directly using `__pydantic_validator__` is considered a stable API? Having a top level function would be nice but really cool would be if one could have a context in which any assignments would be validated.

**Comment by sneakers-the-rat** [comment]
I'd also love to be able to enable validation for a single field or type, e.g. in the case of sqlmodel where validating all assignment would be super expensive, but validating individual fields to apply coercions (e.g. adding timezones to datetimes on loaded objects) would be extremely useful.

Something like this:

```python
UTCDateTime = Annotated[
    datetime, 
    AfterValidator(lambda x: x.replace(tzinfo=UTC)), 
    ValidateAssignment()
]

# or without the `ValidateAssignment()` annotation...

class MyModel(BaseModel):
    utc_datetime: UTCDateTime = Field(validate_assignment=True)
```

I'm not sure how hard this would be, seems like it would just be a matter of adding a field to the field metadata, at a model level treating `validate_assignment` as if it were true by overriding the `__setattr__` method, but then only checking those fields for which it is enabled. would be down to PR this if maintainers would accept it.

**Comment by EytanDn** [comment]
This would be great

**Comment by sneakers-the-rat** [comment]
I had forgotten about this when I raised : https://github.com/pydantic/pydantic/issues/11787

I believe @Viicos has zeroed in on an `on_setattr` callback method set in a `Field` object as the pattern, that issue has some other pitches, but really I would be fine with any impl (and am happy to help)
